zeek/plugin/lib/protocols/zeek_dhcp.spicy

```diff
@@ -0,0 +1,230 @@
+
+        if (option?.request_values) {
+            local values: vector<uint32>;
+
+            for (value in values)
+                values.push_back(cast<uint32>(value));
+
+            param_list.push_back(values);
+        }
+
+        if (option?.client_type) {
+            client_id.push_back(
+                (cast<uint32>(option.client_type),
+                 option.client_identifier.decode(hilti::Charset::ASCII)));
+        }
+
+        if (option?.time_offset)
+            time_offset.push_back(option.time_offset);
+
+        if (option?.routers)
+            for (address in option.routers)
+                routers.push_back(address);
+
+        if (option?.time_servers)
+            for (address in option.time_servers)
+                time_servers.push_back(address);
+
+        if (option?.name_servers)
+            for (address in option.name_servers)
+                name_servers.push_back(address);
+
+        if (option?.domain_name_servers)
+            for (address in option.domain_name_servers)
+                domain_name_servers.push_back(address);
+
+        if (option?.host_name)
+            host_name.push_back(option.host_name);
+
+        if (option?.domain_name)
+            domain_name.push_back(option.domain_name);
+
+        if (option?.forwarding)
+            forwarding.push_back(option.forwarding);
+
+        if (option?.broadcast_address)
+            broadcast_address.push_back(option.broadcast_address);
+
+        if (option?.network_time_protocol_servers)
+            for (address in option.network_time_protocol_servers)
+                network_time_protocol_servers.push_back(address);
+
+        if (option?.vendor_specific_information)
+            vendor_specific_information.push_back(option.vendor_specific_information);
+
+        if (option?.netbios_over_tcpip_name_servers)
+            for (address in option.netbios_over_tcpip_name_servers)
+                netbios_over_tcpip_name_servers.push_back(address);
+
+        if (option?.ip_address_lease_time)
+            ip_address_lease_time.push_back(cast<interval>(option.ip_address_lease_time));
+
+        if (option?.server_identifier)
+            server_identifier.push_back(option.server_identifier);
+
+        if (option?.message)
+            message.push_back(option.message);
+
+        if (option?.maximum_dhcp_message_size)
+            maximum_dhcp_message_size.push_back(cast<uint32>(option.maximum_dhcp_message_size));
+
+        if (option?.renewal_time_value)
+            renewal_time_value.push_back(cast<interval>(option.renewal_time_value));
+
+        if (option?.rebind_time_value)
+            rebind_time_value.push_back(cast<interval>(option.rebind_time_value));
+
+        if (option?.class_identifier)
+            class_identifier.push_back(option.class_identifier);
+
+        if (option?.user_class)
+            user_class.push_back(option.user_class.decode(hilti::Charset::ASCII));
+    }
+
+    return (
```

Comment:
On this Zeek side, is this setting all the `&optional` fields? For example, even if no `routers` are specified, the record field wouldl be set to an empty vector (vs being left unset), right? I just checked, we actually do have the `optional<T>` type  in Spicy already, there are tests for it. If the tuple returned all the values as wrapped in `optional`, it could leave records fields unset where we don't have seen an option it all. But it's ok to leave as it is, seems fine for now.

Reply:
Yes, while this does not fail, it produces incorrect results: when accessing an invalid index, `rt::Vector` will add default-initialized elements to make the access legal; this is definitely not what we want here.

As discussed offline we should leverage the Zeek plugin's support for optional values on both the Spicy and Zeek side for this and use `optional` values in the helper functions added here. I fixed the signatures, but the handling of unset optional fields is currently broken in the plugin, see #401. I pushed updates which at least fix the signatures for now.

Reply:
Filed #415 for other issues and adjusted the code accordingly.